docs/static/js/validator.js

```diff
@@ -0,0 +1,236 @@
+                schema._name = name;
+            }
+        }
+    }
+}
+
+function kebab_to_title(kebab)
+{
+    return kebab.split("-").map(chunk => chunk.charAt(0).toUpperCase() + chunk.substring(1).toLowerCase()).join(" ");
+}
+
+
+function custom_discriminator(propname)
+{
+    function validate_fn(schema, data, parent_schema, data_cxt)
+    {
+        var value = data[propname]
+
+        // Error will be generated by required
+        if ( value === undefined )
+            return true;
+
+        var sub_schema = schema[value];
+        if ( sub_schema === undefined )
+        {
+            validate_fn.errors = [{
+                message: `has unknown '${propname}' value ` + JSON.stringify(value),
+                type: "warning",
+                instancePath: data_cxt.instancePath,
+                parentSchema: parent_schema,
+            }];
+            return false;
+        }
+
+        var validate = this.getSchema(sub_schema.id);
+        if ( !validate(data, data_cxt) )
+        {
+            validate_fn.errors = validate.errors;
+            return false;
+        }
+        return true;
+    }
+
+    return validate_fn;
+}
+
+class Validator
+{
+    constructor(AjvClass, schema_json)
+    {
+        this.schema = schema_json;
+        this.defs = this.schema["$defs"];
+        for ( let [cat, sub_schemas] of Object.entries(this.defs) )
+        {
+            let cat_docs = `/lottie-spec/specs/${cat}/`;
+            let cat_name = kebab_to_title(cat.replace(/s$/, ""));
+            for ( let [obj, sub_schema] of Object.entries(sub_schemas) )
+            {
+                let obj_docs = cat_docs;
+                let obj_name = cat_name;
+                if ( sub_schema.type && obj != "base-gradient" )
+                {
+                    obj_docs += "#" + obj;
+                    obj_name = kebab_to_title(obj);
+                }
+                patch_docs_links(sub_schema, obj_docs, obj_name, obj_name);
+            }
+        }
+        console.log(this.schema);
+
+        let schema_id = this.schema["$id"];
+        this._patch_ty_schema(schema_id, "layers", "all-layers");
+        this._patch_ty_schema(schema_id, "shapes", "all-graphic-elements");
+        for ( let [pname, pschema] of Object.entries(this.defs.properties) )
+        {
+            if ( pname.endsWith("-property") )
+                this._patch_property_schema(pschema, schema_id + "#/$defs/properties/" + pname);
+        }
+
+        this.validator = new AjvClass({
+            allErrors: true,
+            verbose: true,
+            // inlineRefs: false,
+            // strict: false,
+            keywords: [
+                {keyword: ["_docs", "_name", "_docs_name"]},
+                {
+                    keyword: "ty_oneof",
+                    validate: custom_discriminator("ty"),
+                },
+                {
+                    keyword: "prop_oneof",
+                    validate: custom_discriminator("a"),
```

Comment:
What do you think of converting the schema to use if/then/else instead of oneOf for animated properties? That should improve those error messages and would not need any custom logic so would work with any validator.

One thing I'm not sure of is if that would impact the doc generation tools.

Reply:
I think "if" would be easier to maintain as we currently repeat the animated definition in each oneOf.

That is interesting with the tooling support though, do you recall roughly long ago that was? If/then/else was added in draft7 (2018) so maybe it just hadn't had widespread adoption yet?

Reply:
It was like a year or two ago

Reply:
> as we currently repeat the animated definition in each oneOf

Ah my bad, I thought you meant for layers